src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,776 @@
+        /// </summary>
+        ParamArray
+    }
+
+    /// <summary>
+    /// Modes of arguments.
+    /// </summary>
+    public enum ArgumentMode
+    {
+        /// <summary>
+        /// Argument is passed as an input value only.
+        /// </summary>
+        In,
+        /// <summary>
+        /// Argument is passed without an input value and is assigned a value by the time the invoked method returns.
+        /// </summary>
+        Out,
+        /// <summary>
+        /// Argument is passed by reference.
+        /// </summary>
+        Reference
+    }
+
+    /// <summary>
+    /// Represents a named argument.
+    /// </summary>
+    public interface INamedArgument : IArgument
+    {
+        /// <summary>
+        /// Name of the parameter that the argument matches.
+        /// </summary>
+        string Name { get; }
+    }
+
+    /// <summary>
+    /// Represents a reference, which refers to a symbol or an element of a collection.
+    /// </summary>
+    public interface IReference : IExpression
+    {
+        /// <summary>
+        /// Kind of the reference.
+        /// </summary>
+        ReferenceKind ReferenceKind { get; }
+    }
+
+    /// <summary>
+    /// Kinds of references.
+    /// </summary>
+    public enum ReferenceKind
+    {
+        Local,
+        Parameter,
+        Temporary,
+        ArrayElement,
+        PointerIndirection,
+        StaticField,
+        InstanceField,
+        StaticMethod,
+        InstanceMethod,
+        StaticPropertyGet,
+        InstancePropertyGet,
+        StaticPropertySet,
+        InstancePropertySet,
+        StaticProperty,
+        InstanceProperty,
+        ConstantField,
+        LateBoundMember
+    }
+
+    public interface IArrayElementReference : IReference
+    {
+        IExpression ArrayReference { get; }
+        ImmutableArray<IExpression> Indices { get; }
+    }
+
+    public interface IPointerIndirectionReference : IReference
+    {
+        IExpression Pointer { get; }
+    }
+
+    public interface ILocalReference : IReference
+    {
+        ILocalSymbol Local { get; }
+    }
+
+    public interface IParameterReference : IReference
+    {
+        IParameterSymbol Parameter { get; }
+    }
+
+    public interface ITemporaryReference : IReference
+    {
+        TemporaryKind TemporaryKind { get; }
+        IStatement ContainingStatement { get; }
+    }
+
+    /// <summary>
+    /// Kinds of temporary references.
+    /// </summary>
+    public enum TemporaryKind
+    {
+        None,
+
+        /// <summary>
+        /// Temporary created for the step value of a for loop.
+        /// </summary>
+        StepValue,
+        /// <summary>
+        /// Temporary created for the limit value of a for loop.
+        /// </summary>
+        LimitValue
+    }
+
+    public interface IThisReference : IParameterReference
+    {
+        bool IsExplicit { get; }
```

Comment:
It enables higher-fidelity source-to-source transformation, and allows writing at least two analyzers (encourage explicit me/this, and discourage explicit me/this) without recourse to language-specific representations.